libpius/path.py

```diff
@@ -0,0 +1,119 @@
+'''A set of path functions and variables for the PIUS suite.'''
+from __future__ import print_function
+
+import os
+import sys
+import stat
+from os.path import abspath
+import fnmatch
+
+'''This is grabbed right from the python repo under lib/shutil.py'''
+
+def which(cmd, mode=os.F_OK | os.X_OK, path=None):
+    """Given a command, mode, and a PATH string, return the path which
+    conforms to the given mode on the PATH, or None if there is no such
+    file.
+    `mode` defaults to os.F_OK | os.X_OK. `path` defaults to the result
+    of os.environ.get("PATH"), or can be overridden with a custom search
+    path.
+    """
+    # Check that a given file can be accessed with the correct mode.
+    # Additionally check that `file` is not a directory, as on Windows
+    # directories pass the os.access check.
+    def _access_check(fn, mode):
+        return (os.path.exists(fn) and os.access(fn, mode)
+                and not os.path.isdir(fn))
+
+    # If we're given a path with a directory part, look it up directly rather
+    # than referring to PATH directories. This includes checking relative to the
+    # current directory, e.g. ./script
+    if os.path.dirname(cmd):
+        if _access_check(cmd, mode):
+            return cmd
+        return None
+
+    if path is None:
+        path = os.environ.get("PATH", os.defpath)
+    if not path:
+        return None
+    path = path.split(os.pathsep)
+
+    if sys.platform == "win32":
+        # The current directory takes precedence on Windows.
+        if not os.curdir in path:
+            path.insert(0, os.curdir)
+
+        # PATHEXT is necessary to check on Windows.
+        pathext = os.environ.get("PATHEXT", "").split(os.pathsep)
+        # See if the given file matches any of the expected path extensions.
+        # This will allow us to short circuit when given "python.exe".
+        # If it does match, only test that one, otherwise we have to try
+        # others.
+        if any(cmd.lower().endswith(ext.lower()) for ext in pathext):
+            files = [cmd]
+        else:
+            files = [cmd + ext for ext in pathext]
+    else:
+        # On other platforms you don't have things like PATHEXT to tell you
+        # what file suffixes are executable, so just pass on cmd as-is.
+        files = [cmd]
+
+    seen = set()
+    for dir in path:
+        normdir = os.path.normcase(dir)
+        if not normdir in seen:
+            seen.add(normdir)
+            for thefile in files:
+                name = os.path.join(dir, thefile)
+                if _access_check(name, mode):
+                    return name
+    return None
+
+def gpg_test():
+    gpg = which('gpg2')
+    if gpg == "":
+        print("GPG2 could not be found! Is it accessable to $PATH?")
+        sys.exit(1)
+    else:
+        return gpg
+
+def get_home():
+    if sys.platform == "win32":
```

Comment:
This seems right to me....but home is defined on Windows, which means existing windows users will have their configs suddenly not taking effect, right?

Reply:
If you look carefully it says if sys.platform is windows then return os.environ.get APPDATA which is different per user and is technically the users home directory, the else is for unix and linux since they both use HOME

Reply:
I know it's Windows. But I thought windows *also* had a HOME in addition to APPDATA so this would mean existing windows users would have their data in HOME/.piusrc, but we'd be looking for in in APPDATA/.piusrc - I don't want to break backcopmat for people

Reply:
the APPDATA path is the same as all the . dirs and files in unix/linux and contains the settings for how things are viewed by the user and other such settings. Also native windows users will be more familiar with the use of APPDATA as that is where as the name implies all the application data goes.

It is also where gnupg stores their files for windows hosts, which is where we should also store our data files for pius on windows hosts. (symbolic link/shortcut can be made to the users "profile" dir to get easy access.)

Note: What do you mean by backcompat? do you mean on windows hosts or between linux/unix and windows?

Reply:
backcompat: I mean if someone on Windows used `pius` before this PR, put their configs in some place where `pius` was previously looking, then upgrades to the version with this PR, and then `pius` looks in a new place, their experience is bad. I want to make sure we'd have backwards-compatible code to make sure that if users have made config files based on existing behavior, we don't just stop seeing them with the next release.

Today, before this PR, on Windows, we look in `$HOME` - if that isn't defined, then did...
   (1) users define `$HOME` to make it work? If so, will this break them
   (2) the software evaluate `$HOME/.piusrc` as `/.piusrc`? On Windows, does that evaluate to `C:\.piusrc` ? If so, does that mean that users might have put their configs in `C:\.piusrc` ? And if so, this would break them because now we'll be looking in `$APPDATA/.piusrc`, right?

These are the things I'm concerned about - the change in behavior, even though what we're doing today is clearly wrong on Windows, and your patches clearly make the Windows experience better.

So, if one of those is the case, what can we do? Well, we could check _both_ locations for a few releases. Or we could - if we detect a config in the old location - *move* the config to the new location transparently for the user (much like if we detect and old format of `.piusrc` we transparently convert it).

As it stands I don't know the experience of `pius` on Windows, so I can't easily know what existing [bad] behavior of `pius` users may be depending on

Reply:
It wouldn't be possible for someone to use this on windows before with PR seeing as all the paths were hard coded and that the file structure layout is completely different from linux/unix.

Example: linux/unix you use / for the root directory. Windows it is C:\ as the root directory.

Because of this difference the users were unable to use pius on windows unless they were tech and software savvy, many though don't use gpg and pgp on windows and won't have this issue. However the few that do use windows is left out.

As it stands technically there was never windows support for pius before this PR, after it should give windows users access to this software to use. (Note: There is still the issue of the program needing to be compiled in a exe format to be run able which python should be able to do, just need to get the spec file done for it.)

Reply:
Got it. OK then that works for me.